### ISSUE TITLE
Smarter supply converter

### ISSUE THREAD

**Issue by asl97** [minetest-technic/technic#225]
it should adjust the demand base on the load instead of just draining 10000 endlessly.
even if it use a standby demand of 1000 (10% of the max demand in the current code), it would still be better than 10000.

---

where `get_load()` would most probably be named `meta:get_int(to.."_EU_load")`
1:

```
local max_demand = 10000
local input = meta:get_int(from.."_EU_input")
local remain = 0.9
local load = get_load()
local demand
if load/0.9 > max_demand then
    demand = max_demand
else
    demand = load/0.9
end
meta:set_int(from.."_EU_demand", demand)
meta:set_int(from.."_EU_supply", 0)
meta:set_int(to.."_EU_demand", 0)
meta:set_int(to.."_EU_supply", input * remain)
```

2:

```
local max_demand = 10000
local input = meta:get_int(from.."_EU_input")
local SC_demand = 1000
local load = get_load()
if load+SC_demand > max_demand then
    demand = max_demand
else
    demand = load+SC_demand
end
meta:set_int(from.."_EU_demand", demand)
meta:set_int(from.."_EU_supply", 0)
meta:set_int(to.."_EU_demand", 0)
meta:set_int(to.."_EU_supply", input - SC_demand)
```

---

i know it is currently impossible to get the load, a way to get the load would need to be added somehow before it would be possible do to this.

**Comment by asl97** [comment]
i just realize that this might break #56, it's going to get complex.

maybe a check for not full battery would avoid breaking it.

**Comment by thatgraemeguy** [comment]
It's been mentioned before how complex this would be given the current implementation of power networks. Definitely would be nice to have at some point.

**Comment by JDCodeIt** [comment]
I have made some trial code changes that would make the converter flex with demand, but I'm not totally happy with it. It works, though.
- In the switch abm, it stores the total demand and supply in the switch's meta data,
- The entire cable network is also collected when the network is traversed and cached
- The switch position is recorded for each cable in the network
- When the converter runs, it finds the switching station, gets its meta data for total net supply on the upper side and total net demand on the lower, and does a conversion of power types based on an assumed HV=1200V, MV=240V, LV=120V up to the limit of the supply side.
- I do like that a nuclear plant can supply a lot of MV machines - seems reasonable

I don't like:
- that the converter has to search the cable network to find which one it's attached to and therefore where is the switch for that network - every second. Some caching mechanism would be nice, but it needs to know when the network has been adjusted and the cache refreshed.
- when it runs, supply is usually a second behind the net demand, so either under-supplies, or over-supplies in the following second.  Batteries in the network help to smooth out this load fluctuation.

Anyway if anyone is interested, I could fork this and upload the changes for review.

Any other thoughts on the downsides of this implementation?

**Comment by asl97** [comment]
to be honest, i think the power should stay the same

**Comment by HybridDog** [comment]
you can automatically disable it if no power needs to be converted using a piston, mesecons wires and a node detector

**Comment by asl97** [comment]
@HybridDog how would you automatically detect when there is no power needed?

**Comment by HybridDog** [comment]
l would make the node detector test a battery box if it's a full one.

**Comment by asl97** [comment]
i am sure the node detector doesn't have that ability.

**Comment by Ekdohibs** [comment]
@asl97 it can, but this exploits the fact that battery boxes change their node to show the battery level.

**Comment by numberZero** [comment]
What’s about manual control? #288 Not the best way, but seems to work.

**Comment by VanessaE** [comment]
Between the fact that a supply converter uses a percentage and not a fixed amount of EU, and #288 being merged, I think this issue is solved well enough.
